### PR TITLE
Add helpful warning on logger serialization failure

### PR DESCRIPTION
**Pull request by omus**
I've encountered an issue with the following MWE:
```julia
using Distributed, Memento
const LOGGER = getlogger()

addprocs(1)

@everywhere begin
    using Memento

    struct AsyncHandler <: Handler{Union{}, IOBuffer}
        buf::IOBuffer
        channel::Channel{Record}

        function AsyncHandler(buf::IOBuffer)
            channel = Channel{Record}(Inf)
            handler = new(buf, channel)

            task = @async process_record!(handler)
            bind(channel, task)

            return handler
        end
    end

    function process_record!(handler::AsyncHandler)
        while isopen(handler.channel)
            record = take!(handler.channel)
            println(handler.buf, record[:msg])
        end
    end

    function Memento.emit(handler::AsyncHandler, record::Record)
        put!(handler.channel, record)
    end

    push!(getlogger(), AsyncHandler(IOBuffer()))
end

pmap(1:3) do i
    info(LOGGER, i)  # ERROR: cannot serialize a running Task
end
```
Originally I encountered this issue with using the handler in [CloudWatchLogs.jl](https://github.com/invenia/CloudWatchLogs.jl) which also uses an [`@async` task for handling logs](https://github.com/invenia/CloudWatchLogs.jl/blob/29986bc9399bf9c2622589dbfb617e6d687adfa3/src/handler.jl#L42).

The obvious answer is to not use a shared logger across multiple processes but in practise most packages as it'll be slower than using the logger available on each worker. Unfortunately the pattern of the referencing a global logger is pretty common so I expect that this will not be the last time it will be encountered. 

This PR adds in a warning when serialization of a `Logger` fails.

### CODE REVIEW

File: test/handlers.jl [hunk 1]
```diff
@@ -68,6 +68,70 @@
...
+                addprocs(1)
+                @everywhere using Memento
+                # Note: Define AsyncHandler on all processes
+                @everywhere push!(getlogger(), AsyncHandler(IOBuffer()))
+                pmap(1:3) do i
+                    info(LOGGER, i)  # ERROR: cannot serialize a running Task
+                end
+                =#
+                try
+                    serialize(io, handler)
```
Comment (omus) [inline test/handlers.jl]: I need a `@test false` here when no exception is given

File: test/handlers.jl [hunk 2]
```diff
@@ -68,6 +68,72 @@
...
+                    @test false
+                catch e
+                    @test e isa ErrorException
+                    @test e.msg == "cannot serialize a running Task"
+                end
+
+                try
+                    # Emits a warning of: "Logger(AsyncHandler) was unable to be serialized"
+                    serialize(io, logger)
```
Comment (omus) [inline test/handlers.jl]: `@test_warn` doesn't seem to work here as an exception is raised

File: src/loggers.jl [hunk 3]
```diff
@@ -161,6 +161,18 @@ function getpath(logger::Logger)
     return results
 end
 
+function Serialization.serialize(s::AbstractSerializer, logger::Logger)
+    try
+        invoke(serialize, Tuple{AbstractSerializer, Any}, s, logger)
+    catch e
```
Comment (iamed2) [inline src/loggers.jl]: I think you should check for the specific Task serialization error here, and then throw a custom error. If the error is something else, `rethrow` without doing anything.

File: test/handlers.jl (hunk 2 above)
Comment (rofinn) [reply test/handlers.jl]: Shouldn't be an issue anymore if you follow Eric's suggestion and throw a custom error.

File: src/loggers.jl (hunk 3 above)
Comment (omus) [reply src/loggers.jl]: Ok, I was slightly concerned about loosing the original stacktrace

File: test/handlers.jl (hunk 2 above)
Comment (omus) [reply test/handlers.jl]: Sounds good

File: src/loggers.jl (hunk 3 above)
Comment (iamed2) [reply src/loggers.jl]: If you pass your new error to `rethrow` it will keep the old stacktrace! :)

File: src/loggers.jl [hunk 4]
```diff
@@ -161,6 +161,18 @@ function getpath(logger::Logger)
...
+        if e isa ErrorException && e.msg == "cannot serialize a running Task"
+            throw(LoggerSerializationError(logger))
```
Comment (iamed2) [inline src/loggers.jl]:
```suggestion
            rethrow(LoggerSerializationError(logger))
```

File: src/loggers.jl (hunk 3 above)
Comment (omus) [reply src/loggers.jl]: TIL. Should I do that in this case?

Comment (omus) [reply src/loggers.jl]: Answer: yes

**Comment by omus** [comment]
Feel free to squash

**Pull request merged**